### PR TITLE
Race building execution and builder block with cutoff

### LINKED ISSUE #5219: Vc keeps waiting for both builder and engine blocks
**Describe the bug**

This happens on mainnet, bn log:

```
Feb-28 19:45:35.000[chain]         ^[[36mverbose^[[39m: Clock slot slot=5899126
Feb-28 19:45:35.040[rest]            ^[[34mdebug^[[39m: Req req-gssm produceBlindedBlock
Feb-28 19:45:35.129[rest]            ^[[34mdebug^[[39m: Req req-gssn getProposerDuties
Feb-28 19:45:35.129[rest]            ^[[34mdebug^[[39m: Req req-gsso produceBlockV2
Feb-28 19:45:35.317[rest]            ^[[34mdebug^[[39m: Res req-gssn getProposerDuties - 200
Feb-28 19:45:36.539[rest]            ^[[34mdebug^[[39m: Res req-gssm produceBlindedBlock - 200
Feb-28 19:45:44.222[rest]            ^[[34mdebug^[[39m: Res req-gsso produceBlockV2 - 200
```
==> there was 9s delay producing block from bn `produceBlockV2` and vc keeps waiting for that

vc logs

```
Feb-28 19:45:35.001[]                ^[[34mdebug^[[39m: Producing block slot=5899126, validator=0xb88b38d7e613353b29120bd1c34e53db77deeaf03ed30111c273fa6ba61e3d63ab713fd99e090ebbf87492637cef46b8
Feb-28 19:45:35.002[]                ^[[34mdebug^[[39m: HttpClient request routeId=produceBlindedBlock
Feb-28 19:45:35.002[]                ^[[34mdebug^[[39m: HttpClient request routeId=produceBlockV2
Feb-28 19:45:35.642[]                ^[[34mdebug^[[39m: HttpClient response routeId=produceBlockV2
Feb-28 19:45:36.571[]                ^[[34mdebug^[[39m: HttpClient response routeId=produceBlindedBlock
Feb-28 19:45:47.004[]                ^[[34mdebug^[[39m: Request error, retrying  Timeout request
Error: Timeout request
    at HttpClient.requestWithBody (file:///usr/app/packages/api/src/utils/client/httpClient.ts:298:17)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
Feb-28 19:45:47.004[]                ^[[34mdebug^[[39m: Requesting fallback URL routeId=produceBlockV2, baseUrl=http://***:9596, score=10
Feb-28 19:45:59.000[]                ^[[34mdebug^[[39m: HttpClient request routeId=getProposerDuties
Feb-28 19:45:59.005[]                ^[[31merror^[[39m: Failed to produce execution block  Timeout request
Error: Timeout request
    at HttpClient.requestWithBody (file:///usr/app/packages/api/src/utils/client/httpClient.ts:298:17)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
Feb-28 19:45:59.005[]                ^[[34mdebug^[[39m: Produced block slot=5899126, validator=0xb88b38d7e613353b29120bd1c34e53db77deeaf03ed30111c273fa6ba61e3d63ab713fd99e090ebbf87492637cef46b8, source=builder, blockValue=40189560818336940
```

==> vc wait for `produceBlock` on first url then got timed out error after 12s, then try the fallback url and got timed out error after 12s, it's 24s in total

vc got the blinded block early but had to wait for 24s for nothing (supposed to get execution block)

The result is a big delay of block reaching bn and that block became missed

```
beacon-2023-02-28.log:1047384:Feb-28 19:46:00.407[chain]         verbose: Added block to forkchoice slot=5899126, root=0x6e3272efebf0b145a3462d782e95115d605456c5dba5856939c9d594a8f301f8
beacon-2023-02-28.log:1047387:Feb-28 19:46:00.524[chain]         verbose: Block processed slot=5899126, root=0x6e3272efebf0b145a3462d782e95115d605456c5dba5856939c9d594a8f301f8, delaySec=25.5239999294281
```

**Expected behavior**

In this case the blinded block was on time, we should implement some timeout and pick whatever return instead of waiting for both execution block and builder block

### PR DESCRIPTION
**Pull request by g11tech**
We have scenarios where builder api responded with block but execution api took forever to resolve causing an avoidable delay in block proposal (and hence orphaned block)

This PR races building execution and builder block with cutoff of 3 seconds post which the race resolves with whoever wins 

- [x] testcases for `racePromisesWithCutoff` utility
  image

Closes #5219

### CODE REVIEW

**Comment by wemeetagain** [comment]
Can this be done with somehting like:
```ts
function allSettledWithTimeout<T>(promises: Promise<T>, timeoutMs: number) {
  const timerPromise = new Promise((_resolve, reject) => setTimeout(reject, timeoutMs));
  return Promise.allSettled(promises.map((promise) => Promise.race(promise, timerPromise)));
}

**Comment by g11tech** [comment]
> Can this be done with somehting like:
> 
> ```ts
> function allSettledWithTimeout<T>(promises: Promise<T>, timeoutMs: number) {
>   const timerPromise = new Promise((_resolve, reject) => setTimeout(reject, timeoutMs));
>   return Promise.allSettled(promises.map((promise) => Promise.race(promise, timerPromise)));
> }
> ```

problem with this is ( i think) it will force timeout both execution/blinded block at 3 sec for e.g. , but we want to resolve for whoever completes first post cutoff if all promises exceeds cutoff

let me see how i can code it up in a cleaner manner

File: packages/utils/src/promise.ts [hunk 1]
```diff
@@ -24,3 +24,98 @@ export async function callFnWhenAwait<T>(
...
+    if (timeoutObserved) {
+      eventCb(RaceEvent.timeout);
+    } else {
+      eventCb(RaceEvent.pretimeout);
+    }
+  });
+  return mapStatues(promisesStates);
+}
+
+// Some testcases vectors
```
Comment (g11tech) [inline packages/utils/src/promise.ts]: TODO: cleanup

File: tsconfig.build.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "compilerOptions": {
     "target": "es2019",
-    "lib": ["es2020", "esnext.bigint", "es2020.string", "es2020.symbol.wellknown", "dom"],
+    "lib": ["es2020", "esnext.bigint", "es2020.string", "es2020.symbol.wellknown", "dom", "es2021.promise"],
```
Comment (g11tech) [inline tsconfig.build.json:4]: required for Promise.any

**Comment by g11tech** [comment]
@wemeetagain i have updated the racefn using the inbuild primitives, let me know how it looks now

File: packages/validator/src/services/block.ts [hunk 3]
```diff
@@ -11,13 +11,20 @@ import {ValidatorStore, BuilderSelection} from "./validatorStore.js";
 import {BlockDutiesService, GENESIS_SLOT} from "./blockDuties.js";
 
 const ETH_TO_WEI = BigInt("1000000000000000000");
+const BLOCK_PRODUCTION_RACE_CUTOFF_MS = 3_000;
+const BLOCK_PRODUCTION_RACE_TIMEOUT_MS = 12_000;
```
Comment (g11tech) [inline packages/validator/src/services/block.ts:20]: are these values optimal (or good enough to being with)?

Comment (nflaig) [reply packages/validator/src/services/block.ts:20]: do you know how other CL clients handle this and what cutoff they use?

File: packages/beacon-node/test/scripts/el-interop/mergemock/post-merge.sh [hunk 4]
```diff
@@ -6,4 +6,4 @@ currentDir=$(pwd)
 . $scriptDir/common-setup.sh
 
 # if we don't provide any datadir merge mock stores data in memory which is fine by us
-docker run --rm -u $(id -u ${USER}):$(id -g ${USER}) --name custom-execution --network host -v $currentDir/$DATA_DIR/genesis.json:/usr/app/genesis.json -v $currentDir/$DATA_DIR/jwtsecret:/usr/app/jwt.hex $EL_BINARY_DIR relay --listen-addr 127.0.0.1:$ETH_PORT --engine-listen-addr 127.0.0.1:$ENGINE_PORT --log.level debug --genesis-validators-root 0x3e8bd71d9925794b4f5e8623e15094ea6edc0fd206e3551e13dd2d10e08fbaba
+docker run --rm -u $(id -u ${USER}):$(id -g ${USER}) --name custom-execution --network host -v $currentDir/$DATA_DIR/genesis.json:/usr/app/genesis.json -v $currentDir/$DATA_DIR/jwtsecret:/usr/app/jwt.hex $EL_BINARY_DIR relay --listen-addr 0.0.0.0:$ETH_PORT --engine-listen-addr 0.0.0.0:$ENGINE_PORT --log.level debug --genesis-validators-root 0x3e8bd71d9925794b4f5e8623e15094ea6edc0fd206e3551e13dd2d10e08fbaba
```
Comment (twoeths) [inline packages/beacon-node/test/scripts/el-interop/mergemock/post-merge.sh]: do we need this change? also the one in `mergemock.test.ts`

File: packages/utils/src/promise.ts [hunk 5]
```diff
@@ -24,3 +24,110 @@ export async function callFnWhenAwait<T>(
...
+  rejected,
+  pending,
+}
+
+type PromiseState<T> =
+  | {status: PromiseStatus.resolved; value: T}
+  | {status: PromiseStatus.rejected; value: Error}
+  | {status: PromiseStatus.pending; value: null};
+
+function mapStatuesToResponses<T>(promisesStates: PromiseState<T>[]): (Error | T)[] {
```
Comment (twoeths) [inline packages/utils/src/promise.ts]: typo?

File: packages/validator/src/services/block.ts (hunk 3 above)
Comment (twoeths) [reply packages/validator/src/services/block.ts:20]: these values make sense to me

File: packages/beacon-node/test/scripts/el-interop/mergemock/post-merge.sh (hunk 4 above)
Comment (g11tech) [reply packages/beacon-node/test/scripts/el-interop/mergemock/post-merge.sh]: not essential for this PR, i can bundle in a separate PR :+1:

File: packages/utils/src/promise.ts (hunk 5 above)
Comment (g11tech) [reply packages/utils/src/promise.ts]: ohh yea, thanks for spotting :pray:

File: packages/utils/src/promise.ts [hunk 6]
```diff
@@ -24,3 +24,118 @@ export async function callFnWhenAwait<T>(
...
+    }
+  });
+}
+
+export enum RaceEvent {
+  /** all reject/resolve before cutoff */
+  precutoff = "precutoff-return",
+  /** cutoff reached as some were pending till cutoff **/
+  cutoff = "cutoff-reached",
+  /** atleast one resolved till cutoff so no race required */
```
Comment (nflaig) [inline packages/utils/src/promise.ts:57]:
```suggestion
  /** at least one resolved till cutoff so no race required */
```

File: packages/utils/src/promise.ts [hunk 7]
```diff
@@ -24,3 +24,118 @@ export async function callFnWhenAwait<T>(
...
+      return mapStatusesToResponses(promisesStates);
+    } else {
+      eventCb(RaceEvent.cutoff, Date.now() - startTime);
+    }
+  } else {
+    eventCb(RaceEvent.precutoff, Date.now() - startTime);
+    return mapStatusesToResponses(promisesStates);
+  }
+
+  // Post deadline resolve with any of the promise or all rejected before timeout
```
Comment (nflaig) [inline packages/utils/src/promise.ts:129]: deadline is the cutoff here, right?

```suggestion
  // Post cutoff resolve with any of the promise or all rejected before timeout
```

**Review by nflaig — APPROVED** [review]
looks really clean and well tested!

File: packages/utils/test/unit/promiserace.test.ts [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+  const pretimeout = RaceEvent.pretimeout;
+  const timeout = RaceEvent.timeout;
+  const resolved = RaceEvent.resolved;
+  const rejected = RaceEvent.rejected;
+
+  // Second item in testcase row i.e. array of numbers represent delay at which promise to be resolved
+  // or rejected (-ve number)
+  // Third item in testcase row is the expected value or error message in string
+  // Last item is expected events
+  const testcases: [string, number[], (string | Error)[], RaceEvent[]][] = [
```
Comment (twoeths) [inline packages/utils/test/unit/promiserace.test.ts:30]: thanks for a bunch of test cases 👍

**Review by twoeths — APPROVED** [review]

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.6.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.6.0) :tada:

**Pull request merged**